SignalServiceKit/src/Messages/OWSMessageSender.m

```diff
@@ -168,144 +107,98 @@ @implementation OWSSendMessageOperation
 
     return self;
 }
 
-#pragma mark - NSOperation overrides
+#pragma mark - OWSOperation overrides
 
-- (BOOL)isExecuting
+- (nullable NSError *)checkForPreconditionError
 {
-    return self.operationState == OWSSendMessageOperationStateExecuting;
-}
+    for (NSOperation *dependency in self.dependencies) {
+        if (![dependency isKindOfClass:[OWSOperation class]]) {
+            NSString *errorDescription =
+                [NSString stringWithFormat:@"%@ unknown dependency: %@", self.logTag, dependency.class];
+            NSError *assertionError = OWSErrorMakeAssertionError(errorDescription);
+            return assertionError;
+        }
 
-- (BOOL)isFinished
-{
-    return self.operationState == OWSSendMessageOperationStateFinished;
-}
+        OWSOperation *upload = (OWSOperation *)dependency;
 
-- (void)start
-{
-    [self willChangeValueForKey:OWSSendMessageOperationKeyIsExecuting];
-    self.operationState = OWSSendMessageOperationStateExecuting;
-    [self didChangeValueForKey:OWSSendMessageOperationKeyIsExecuting];
-    [self main];
-}
+        // Cannot proceed if dependency failed - surface the dependency's error.
+        NSError *_Nullable dependencyError = upload.failingError;
+        if (dependencyError) {
+            return dependencyError;
+        }
+    }
 
-- (void)main
-{
-    [self tryWithRemainingRetries:OWSSendMessageOperationMaxRetries];
-}
+    // Sanity check preconditions
+    if (self.message.hasAttachments) {
+        [self.dbConnection readWithBlock:^(YapDatabaseReadTransaction *_Nonnull transaction) {
+            TSAttachmentStream *attachmentStream
+                = (TSAttachmentStream *)[self.message attachmentWithTransaction:transaction];
+            OWSAssert(attachmentStream);
+            OWSAssert([attachmentStream isKindOfClass:[TSAttachmentStream class]]);
+            OWSAssert(attachmentStream.serverId);
+            OWSAssert(attachmentStream.isUploaded);
+        }];
+    }
 
-#pragma mark - methods
+    return nil;
+}
 
-- (void)tryWithRemainingRetries:(NSUInteger)remainingRetries
+- (void)run
 {
     // If the message has been deleted, abort send.
     if (self.message.shouldBeSaved && ![TSOutgoingMessage fetchObjectWithUniqueID:self.message.uniqueId]) {
         DDLogInfo(@"%@ aborting message send; message deleted.", self.logTag);
         NSError *error = OWSErrorWithCodeDescription(
             OWSErrorCodeMessageDeletedBeforeSent, @"Message was deleted before it could be sent.");
-        self.failureHandler(error);
+        error.isFatal = YES;
+        [self reportError:error];
         return;
     }
 
-    // Use this flag to ensure a given operation only succeeds or fails once.
-    __block BOOL onceFlag = NO;
-    RetryableFailureHandler retryableFailureHandler = ^(NSError *_Nonnull error) {
-        DDLogInfo(@"%@ Sending failed. Remaining retries: %lu", self.logTag, (unsigned long)remainingRetries);
-
-        OWSAssert(!onceFlag);
-        onceFlag = YES;
```

Comment:
This once flag is captured in the synchronized assertion in the super class, which asserts that markForCompletion is only ever called once.